File: swift/Sources/AlloyTime/ZoneFormat.swift
import Foundation

/// Zone-derived display strings. Mirrored twin of `zone-format.ts`.
public enum ZoneFormat {
    /// Sign + hours, with ":mm" only when the zone is off a whole hour.
    /// Uses U+2212 MINUS for negatives, matching the apps.
    public static func compactOffset(_ date: Date, timeZone: TimeZone) -> String {
        let minutes = timeZone.secondsFromGMT(for: date) / 60
        let sign = minutes < 0 ? "\u{2212}" : "+"
        let abs = Swift.abs(minutes)
        let h = abs / 60, m = abs % 60
        return m == 0 ? "\(sign)\(h)" : "\(sign)\(h):\(String(format: "%02d", m))"
    }

    /// String-id variant (web-mirrored signature); unresolvable ids read +0,
    /// matching the web's bare-"GMT" fallback.
    public static func compactOffset(_ date: Date, zone id: String) -> String {
        compactOffset(date, timeZone: ZoneCatalog.resolve(id) ?? TimeZone(secondsFromGMT: 0)!)
    }

    /// "GMT+05:30" / "GMT−08:00" (U+2212), the web's longOffset rendering.
    public static func gmtOffset(_ date: Date, timeZone: TimeZone) -> String {
        let minutes = timeZone.secondsFromGMT(for: date) / 60
        let sign = minutes < 0 ? "\u{2212}" : "+"
        let abs = Swift.abs(minutes)
        return String(format: "GMT%@%02d:%02d", sign, abs / 60, abs % 60)
    }

    /// City label from an IANA id: last path segment, underscores spaced,
    /// uppercased. `abbreviate` collapses multi-word to initials, single word to
    /// first three letters. Fixed-offset ids ("+05:30") have no city.
    public static func zoneCity(_ ianaId: String, abbreviate: Bool) -> String {
        if ianaId.range(of: "^[+\u{2212}-]\\d", options: .regularExpression) != nil { return "" }
        let city = (ianaId.split(separator: "/").last.map(String.init) ?? ianaId)
            .replacingOccurrences(of: "_", with: " ")
        if !abbreviate { return city.uppercased() }
        let words = city.split(whereSeparator: { $0 == " " || $0 == "-" }).map(String.init)
        let label = words.count > 1 ? words.map { String($0.prefix(1)) }.joined()
                                    : String(city.prefix(3))
        return label.uppercased()
    }
}

File: web/packages/alloy-time/src/zone-format.ts
import { zoneOffsetMinutes } from './zone-catalog';

export function gmtOffset(date: Date, timeZone: string): string {
  const name =
    new Intl.DateTimeFormat('en-US', { timeZone, timeZoneName: 'longOffset' })
      .formatToParts(date)
      .find((part) => part.type === 'timeZoneName')?.value ?? 'GMT';
  const normalized = name === 'GMT' ? 'GMT+00:00' : name;
  return normalized.replace('-', '−');
}

// Compact UTC offset for the date row's globe badge: sign + hours, appending
// ":mm" only when the zone isn't on a whole hour ("−7", "+9", "+5:30", "+0").
// The globe icon stands in for the "GMT" prefix. U+2212 minus, matching the app.
export function compactOffset(date: Date, timeZone: string): string {
  const min = zoneOffsetMinutes(timeZone, date);
  const sign = min < 0 ? '−' : '+';
  const abs = Math.abs(min);
  const hours = Math.floor(abs / 60);
  const minutes = abs % 60;
  return minutes === 0 ? `${sign}${hours}` : `${sign}${hours}:${String(minutes).padStart(2, '0')}`;
}

// City label derived from an IANA zone id: the last path segment, underscores
// spaced out and uppercased ("America/Los_Angeles" -> "LOS ANGELES"). When
// `abbreviate` is true (a country flag already supplies the locale), collapse a
// multi-word city to its initials, or a single word to its first three letters
// ("LOS ANGELES" -> "LA", "London" -> "LON", "UTC" -> "UTC").
export function zoneCity(timeZone: string, abbreviate: boolean): string {
  // Fixed-offset zones ('+05:30', '−08:00') have no city — the globe offset says it.
  if (/^[+−-]\d/.test(timeZone)) return '';
  const city = (timeZone.split('/').pop() ?? timeZone).replace(/_/g, ' ');
  if (!abbreviate) return city.toUpperCase();
  const words = city.split(/[\s-]+/).filter(Boolean);
  const label = words.length > 1 ? words.map((w) => w[0]).join('') : city.slice(0, 3);
  return label.toUpperCase();
}
